### PR TITLE
Add ProfileProvider class to support read profile config from PROFILE_DB.

### PR DESCRIPTION
**Pull request by liuh-80**
#### Why I did it
Add default value and profile support to swss-swss-common.

#### How I did it
Add ProfileProvider class to support read profile config from PROFILE_DB.

#### How to verify it
Add new UT.
Pass all existing UT and E2E test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
Add ProfileProvider class to support read profile config from PROFILE_DB.

### CODE REVIEW

**Comment by liuh-80** [comment]
This PR depends on another PR to pass UT: https://github.com/sonic-net/sonic-swss-common/pull/681
The reason is UT need modify redis config to create a new UT, which is part of azure pipeline change.

File: .azure-pipelines/build-sairedis-template.yml [hunk 1]
```diff
@@ -82,12 +86,31 @@ jobs:
       sudo sed -ri 's/^# unixsocket/unixsocket/' /etc/redis/redis.conf
       sudo sed -ri 's/^unixsocketperm .../unixsocketperm 777/' /etc/redis/redis.conf
       sudo sed -ri 's/redis-server.sock/redis.sock/' /etc/redis/redis.conf
+      sudo sed -ri 's/^databases .*/databases 17/' /etc/redis/redis.conf
```
Comment (liuh-80) [inline .azure-pipelines/build-sairedis-template.yml:89]: This line will add new DB for UT.

All other pipeline change are from https://github.com/sonic-net/sonic-swss-common/pull/681

File: common/dbconnector.cpp [hunk 2]
```diff
@@ -905,3 +905,26 @@ void DBConnector::del(const std::vector<std::string>& keys)
 
     pipe.flush();
 }
+
+map<string, map<string, map<string, string>>> DBConnector::getall()
```
Comment (qiluo-msft) [inline common/dbconnector.cpp:909]: >getall [](http://example.com/codeflow?start=59&length=6)

Move pure refactoring into a separate PR. The benefit is easy to revert or cherry-pick smaller PRs.

File: tests/redis_multi_db_ut_config/database_config0.json [hunk 3]
```diff
@@ -86,6 +86,11 @@
             "id" : 15,
             "separator": ":",
             "instance" : "redis"
+        },
+        "PROFILE_DB" : {
+            "id" : 16,
```
Comment (qiluo-msft) [inline tests/redis_multi_db_ut_config/database_config0.json:91]: >16 [](http://example.com/codeflow?start=19&length=2)

Possible to use 12?

**Comment by qiluo-msft** [comment]
>         "LOGLEVEL_DB" : {

There is a ongoing PR to remove this DB. Possible to use it?

---
Refers to: tests/redis_multi_db_ut_config/database_config1.json:25 in 9b60524. [](commit_id = 9b60524cc312df68a6feed4fb3e0fec60d63d1fb, deletion_comment = False)

File: .azure-pipelines/build-template.yml [hunk 4]
```diff
@@ -134,6 +134,7 @@ jobs:
         sudo sed -ri 's/^# unixsocket/unixsocket/' /etc/redis/redis.conf
         sudo sed -ri 's/^unixsocketperm .../unixsocketperm 777/' /etc/redis/redis.conf
         sudo sed -ri 's/redis-server.sock/redis.sock/' /etc/redis/redis.conf
+        sudo sed -ri 's/^databases .*/databases 17/' /etc/redis/redis.conf
```
Comment (qiluo-msft) [inline .azure-pipelines/build-template.yml:137]: >17 [](http://example.com/codeflow?start=48&length=2)

If this is the final decision, suggest to put serious effort into coordinate PRs in multiple repoes, and make it easy tunable in future.
What is the maximal value? Possible to double if we really need?

File: tests/redis_multi_db_ut_config/database_config0.json (hunk 3 above)
Comment (liuh-80) [reply tests/redis_multi_db_ut_config/database_config0.json]: Fixed, change to use 12

File: .azure-pipelines/build-template.yml (hunk 4 above)
Comment (liuh-80) [reply .azure-pipelines/build-template.yml]: Fixed, remove this line because DB change to use 12.
There is no limit for the max DB count, we can change to any int value.

Will check and fix other repos before this PR merge.

For DB count in sonic image, it's calculated by DBs in database_config.json:
https://github.com/sonic-net/sonic-buildimage/pull/11781

**Comment by liuh-80** [comment]
> > ```
> >     "LOGLEVEL_DB" : {
> > ```
> 
> There is a ongoing PR to remove this DB. Possible to use it?
> 
> Refers to: tests/redis_multi_db_ut_config/database_config1.json:25 in [9b60524](https://github.com/sonic-net/sonic-swss-common/commit/9b60524cc312df68a6feed4fb3e0fec60d63d1fb). [](commit_id = [9b60524](https://github.com/sonic-net/sonic-swss-common/commit/9b60524cc312df68a6feed4fb3e0fec60d63d1fb), deletion_comment = False)

Chage to use 12, so not necessary use ID for this DB

File: tests/redis_multi_db_ut_config/database_config0.json (hunk 3 above)
Comment (liuh-80) [reply tests/redis_multi_db_ut_config/database_config0.json]: DB 12 used by CHASSIS_APP_DB in common/database_config.json, will check if can use LOGLEVEL_DB

File: common/dbconnector.cpp (hunk 2 above)
Comment (liuh-80) [reply common/dbconnector.cpp:909]: Fixed, code change move to this PR: https://github.com/sonic-net/sonic-swss-common/pull/692

File: common/configdb.cpp [hunk 5]
```diff
@@ -244,24 +244,7 @@ void ConfigDBConnector_Native::mod_config(const map<string, map<string, map<stri
 map<string, map<string, map<string, string>>> ConfigDBConnector_Native::get_config()
 {
```
Comment (liuh-80) [inline common/configdb.cpp:245]: File change in configdb.cpp dbconnector.cpp and dbconnector.h are from another PR:
https://github.com/sonic-net/sonic-swss-common/pull/692

File: tests/redis_multi_db_ut_config/database_config0.json (hunk 3 above)
Comment (liuh-80) [reply tests/redis_multi_db_ut_config/database_config0.json]: Fixed, for UT we use DB 12, and in sonic image we still use DB 15, which configured in database_config.json:
https://github.com/sonic-net/sonic-buildimage/pull/11781

File: common/database_config.json [hunk 6]
```diff
@@ -91,6 +91,11 @@
             "id" : 14,
             "separator": ":",
             "instance" : "redis"
```
[the remark this answers is not in the document]
Comment (liuh-80) [reply common/database_config.json]: Fixed

**Review by qiluo-msft — DISMISSED** [review]

**Comment by liuh-80** [comment]
Will merge after this validation pass: https://github.com/sonic-net/sonic-buildimage/pull/10575

**Comment by liuh-80** [comment]
Will merge this PR after the validation PR passed: https://github.com/sonic-net/sonic-buildimage/pull/10575

**Comment by qiluo-msft** [comment]
Could you check build failures?

**Comment by liuh-80** [comment]
> Could you check build failures?
The build failure caused by code change in sairedis, following PR will fix it:
https://github.com/sonic-net/sonic-swss-common/pull/747

**Review by qiluo-msft — APPROVED** [review]
